packages/client/src/objectSet/ObjectSet.test.ts

```diff
@@ -436,6 +437,260 @@ describe("ObjectSet", () => {
+      client(Employee).withProperties({
+        "derivedPropertyName": (base) =>
+          // @ts-expect-error
+          base.aggregate("employeeId:exactDistinct"),
+      });
+
+      client(Employee).withProperties({
+        "derivedPropertyName": (base) =>
+          // @ts-expect-error
+          base.selectProperty("employeeId"),
+      });
+    });
+
+    it("does not allow selectProperty when a many link was selected at any point", () => {
+      client(Employee).withProperties({
+        "derivedPropertyName": (base) => {
+          // @ts-expect-error
+          base.pivotTo("peeps").selectProperty("employeeId");
+
+          // @ts-expect-error
+          base.pivotTo("lead").pivotTo("peeps").selectProperty("employeeId");
+
+          return base.pivotTo("lead").selectProperty("employeeId");
+        },
+      });
+    });
+
+    // Executed code fails since we're providing bad strings to the function
+    it.fails("correctly narrows types of aggregate function", () => {
+      client(Employee).withProperties({
+        "derivedPropertyName": (base) => {
+          // @ts-expect-error
+          base.pivotTo("lead").aggregate("notAProperty:sum");
+
+          // @ts-expect-error
+          base.pivotTo("lead").aggregate(":avg");
+
+          // @ts-expect-error
+          base.pivotTo("lead").aggregate("employeeId:notAnOp");
+
+          base.pivotTo("lead").aggregate("employeeId:collectList");
+
+          return base.pivotTo("lead").aggregate("employeeId:sum");
+        },
+      });
+    });
+
+    // Executed code fails since we're providing bad strings to the function
+    it("correctly narrows types of options for aggregate functions", () => {
+      client(Employee).withProperties({
+        "derivedPropertyName": (base) => {
+          // @ts-expect-error
+          base.pivotTo("lead").aggregate("employeeId:approximateDistinct", {
+            limit: 1,
+          });
+
+          base.pivotTo("lead").aggregate("employeeId:collectList", {
+            limit: 1,
+          });
+
+          base.pivotTo("lead").aggregate("employeeId:collectSet", { limit: 1 });
+
+          base.pivotTo("lead").aggregate("employeeId:collectList", {
+            // @ts-expect-error
+            percentile: 1,
+          });
+
+          return base.pivotTo("lead").aggregate(
+            "employeeId:approximatePercentile",
+            { percentile: 0.5 },
+          );
+        },
+      });
+    });
+
+    it("correctly narrows types of selectProperty function", () => {
+      client(Employee).withProperties({
+        "derivedPropertyName": (base) => {
+          // @ts-expect-error
+          base.pivotTo("lead").selectProperty("notAProperty");
+
+          return base.pivotTo("lead").selectProperty("employeeStatus");
+        },
+      });
+    });
+
+    it("propagates derived property type to future object set operations with correct types", () => {
```

Comment:
This test and others with no clause are just for us to ensure that we're typing things correctly.